testing/pr-testing-guidelines-and-checklists.md

```diff
@@ -0,0 +1,149 @@
+- to catch regressions early
+- to maintain a good quality of incoming code
+
+## Terms
+
+* relevant repository - repository that could influence ship-able artifacts by it's contents in a bad way. If we really don't want to follow these guidelines, just claim that the repository is currently irrelevant for our project.
+  
+## Proposed Solution
+
+These proposals only concern relevant repositories.
+
+### Repository should have a mandatory review by other person
+
+This is easily enabled in GitHub. It mandates that at least one person gives a PR a green check-mark.
+
+There will be some decisions that we don't want to automate on purpose, for example, we will probably display code-coverage prominently,
+but we won't have any automatically enforced that would stop PR from getting merged due to low coverage.
+
+In the future, we might have optional integration tests triggered by comments or specific tags on the PR. This would be hard to enforce automatically,
+so mandating those would fall as a responsibility of the reviewer.
+
+There could be a template with a checklist delivered by a .github/PULL_REQUEST_TEMPLATE.md
+Checklist would include:
+
+* Read through the code and ask about any unclear code
+* Try to checkout and run the relevant code
+* Evaluate if the new code is sufficiently tested, potentially consulting the code-coverage
+* Evaluate if there are any follow-up tasks to be logged
+* more?
+
+### All of the automated testing should finish in reasonable time, defined as in less than 10 minutes
+
+Looking at our previous PR checking attempts, even fast checks take around 2 minutes, due to inconsistencies in time of spawning our on-demand slaves.
+
+Our longest PR builds can take around 20 minutes.
+
+Any time, there is PR check that consistently hits the 10 minute run-time, it should become an issue in our backlog to be solved.
+
+### There should be a consistent and enforced style for our code
+
+Maintaining a repository where multiple developers collaborate should be helped by ensuring consistent style.
+While this often leads to a lot of bike-shedding, and changing the style after it has been applied can be tedious,
+but the style enforcement still seems to be quite useful, especially for new collaborators.
+
+For configuration files and dynamic languages, style checking can often find basic problems that would be discovered in compiled languages with type-checker.
+
+We could experiment even with automatic style formatting, such as:
+* gofmt
+* grunt-jsbeautifier
+
+We should enforce the style for:
+* go, probably with golint
+* javascript, probably with eslint, modeling after openshift-team's style guide
+* ansible, if some check exists
+* bash (proposed by Pavel Sturc)
+
+### If there is a build-able artifact, it should be built in reasonable subset of configurations
+
+This mostly goes for binaries, such as our go-based cli tool, or compilable libraries for our mobile devices.
```

Comment:
by this do you mean build it for windows, mac and linux?

Reply:
I am not sure. 

On one hand there is a great deal of utility even if we compile for just one target, i.e. type checker gets run.

Depends on our trust in the portability of the code. I think it might be fine to just build for linux (as that is what is easily accessible, are there cross compilation targets from linux to win/osx?) 

If we think there is use in having windows/osx builds on every PR (it is probable that most of our customers would use our tools from win/osx), then yes, we should do it.

Reply:
yes it is very easy to compile for different arch in go you should only need to set the env vars:
``` GOOS and GOARCH```

Reply:
Awesome, in that case we should definitely do that.

Do you think we should run the tests on multiple arch as well? That might be the tricky part (I don't really want to maintain windows slaves unless necesary )

Reply:
No I see no need for that